pkg/testutil/testutil.go

```diff
@@ -787,6 +787,9 @@ func newBase(t *testing.T, ns string, ipv6Compatible bool, kubernetesCompatible
 	} else if !base.EnableKubernetes && base.KubernetesCompatible {
 		t.Skip("runner skips Kubernetes compatible tests in the non-Kubernetes environment")
 	}
+	if !GetFlakyEnvironment() && !GetEnableKubernetes() && !GetEnableIPv6() {
+		t.Skip("legacy tests are considered flaky by default and are skipped unless in the flaky environment")
```

Comment:
This sounds like overestimating 🤔

Now those tests are skipped on the CI?

Reply:
Nothing changes for the pre-existing tests - they will still be retried, and will not be skipped.

The difference is that the new tests now run first, in a separate step of the job, without retries.

Reply:
`./test-integration.sh -test.only-flaky`:
- does run with retries, and includes all legacy tests, and all new tests that have been flagged flaky explicitly

`./test-integration.sh`:
- default behavior, runs only the non-flaky tests

The CI runs both, as successive steps of the test job.

Reply:
And yes, we need the github actions PR once this one will be merged.

Reply:
What about adding a flag like `-test.skip-legacy` instead?

Not every legacy test should be assumed to be flaky?

Reply:
> What about adding a flag like `-test.skip-legacy` instead?
> 
> Not every legacy test should be assumed to be flaky?

But this is what we are doing right now: any existing is treated as flaky and will be retried if it fails.

This PR does not change anything for that.

Reply:
> > What about adding a flag like `-test.skip-legacy` instead?
> > Not every legacy test should be assumed to be flaky?
> 
> But this is what we are doing right now: any existing test is treated as flaky and will be retried if it fails.
> 
> This PR does not change anything for that.

Current situation:
- every test is treated as flaky and will be retried multiple times until it passes

With these two PRs:
- pre-existing tests and new tests marked "isflaky" are treated exactly the same as previously and will be retried if they fail - this is what is called "flaky"
- new tests are only run once and are not retried - this is what is called "non-flaky"

Hope that clarifies.

Reply:
How do we distinguish between a `pre-existing` tests and `new` tests?

Do we have many flaky tests? If not I'd prefer assuming all tests (regardless new or pre-existing) are not flaky, and mark individual ones as flaky explicitly.

Reply:
> Do we have many flaky tests? 

As stated previously: we currently re-run all tests multiple times if they fail, effectively treating _every_ test as flaky.
This is the current situation, which is obviously hiding all errors.

I tried multiple times to disable retries before:
- https://github.com/containerd/nerdctl/issues/3303
- https://github.com/containerd/nerdctl/pull/3400
- https://github.com/containerd/nerdctl/pull/3189

> If not I'd prefer assuming all tests (regardless new or pre-existing) are not flaky, and mark individual ones as flaky explicitly.

I do not see how we can do that _from the current situation_, which is as stated to treat  everything as flaky.

Reply:
> Do we have many flaky tests? If not I'd prefer assuming all tests (regardless new or pre-existing) are not flaky, and mark individual ones as flaky explicitly.

I offered this before in https://github.com/containerd/nerdctl/issues/3303 

That would mean the CI would be very finicky and failing for weeks (or months, depending on time invested in this) and we would have to go through a lot of pain to get it back to any form of green.

Furthermore - and this is one of the reasons I started the test rewrite effort, a significant fraction of the failures are problems with the test, not with the code.

At least with the rewrite, I know that the problems are with the code.

Hope that clarifies.